### PR TITLE
refactor: next version of psbt ui

### LINKED ISSUE #3440: Improve PSBT UX
Cleanly list inputs vs. outputs in a way similar to post conditions for Stacks transactions

### PR DESCRIPTION
**Pull request by fbwoolf**
> Try out this version of the Hiro Wallet - download [extension builds](https://github.com/hirosystems/wallet/actions/runs/4691067531).

Hoping I have this right now, been a long exploration/learning process. ChatGPT is now my new best friend. I'm sure there will be edge cases, but this is working from tests with both ME and Gamma.

*I believe I still need to implement a static footer at some point

Screen Shot 2023-04-09 at 11 42 31 AM

Screen Shot 2023-04-09 at 11 41 03 AM

Note: I think we should still show the value here rather than the inscription number? That would be an additional query.
Screen Shot 2023-04-09 at 11 58 45 AM

Screen Shot 2023-04-09 at 11 41 33 AM

Screen Shot 2023-04-09 at 11 41 48 AM

### CODE REVIEW

**Comment by fbwoolf** [comment]
@markmhx @mica000 let me know what you think here.

EDIT: I feel like we might need a display for when our logic can't match up the inputs/outputs and instructs the user to check the advanced view? Or, maybe we force that view? I have a feeling apps might pass us psbts that aren't nec formatted correctly?

**Comment by kyranjamie** [comment]
image

File: src/app/pages/psbt-request/components/psbt-decoded-request/psbt-decoded-request-views/match-inputs-and-outputs.ts [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+    // Iterate through the available inputs and add them until the output is fulfilled
+    for (let i = 0; i < availableInputs.length; i++) {
+      const input = availableInputs[i];
+
+      if (remainingInputValue === 0) remainingInputValue = input.value;
+
+      // Handle if the unsigned input at index has an inscription
+      if (availableInputs[i].unsignedUtxo?.inscriptions) {
+        // Make sure this is an inscription transfer value
+        if (remainingInputValue === 10000 || remainingOutputValue === 10000) {
```
Comment (alter-eggo) [inline src/app/pages/psbt-request/components/psbt-decoded-request/psbt-decoded-request-views/match-inputs-and-outputs.ts]: hm, why this value is `10000`?

File: src/app/pages/psbt-request/components/psbt-decoded-request/psbt-input-output-pair/psbt-inscription.tsx [hunk 2]
```diff
@@ -0,0 +1,12 @@
+import { InscriptionPreview } from '@app/components/inscription-preview-card/components/inscription-preview';
+import { useInscription } from '@app/query/bitcoin/ordinals/inscription.hooks';
+
+export function PsbtInscription(props: { path: string }) {
+  const { path } = props;
+  const { isLoading, isError, data: inscription } = useInscription(path);
+
+  if (isLoading) return null;
```
Comment (alter-eggo) [inline src/app/pages/psbt-request/components/psbt-decoded-request/psbt-input-output-pair/psbt-inscription.tsx]: `isLoading || isError` ?

File: src/app/pages/psbt-request/components/psbt-decoded-request/psbt-decoded-request-views/match-inputs-and-outputs.ts (hunk 1 above)
Comment (fbwoolf) [reply src/app/pages/psbt-request/components/psbt-decoded-request/psbt-decoded-request-views/match-inputs-and-outputs.ts]: This is the transfer value of an inscription.

File: src/app/pages/psbt-request/components/psbt-decoded-request/psbt-input-output-pair/psbt-inscription.tsx (hunk 2 above)
Comment (fbwoolf) [reply src/app/pages/psbt-request/components/psbt-decoded-request/psbt-input-output-pair/psbt-inscription.tsx]: Yeah, was separating to possibly handle differently, but thanks for reminding me.

File: src/app/pages/psbt-request/components/psbt-decoded-request/psbt-decoded-request-views/match-inputs-and-outputs.ts (hunk 1 above)
Comment (alter-eggo) [reply src/app/pages/psbt-request/components/psbt-decoded-request/psbt-decoded-request-views/match-inputs-and-outputs.ts]: may be create variable for it?

File: src/app/pages/psbt-request/components/psbt-decoded-request/psbt-decoded-request-views/build-psbt-input-for-ui.ts [hunk 3]
```diff
@@ -0,0 +1,46 @@
+import { BitcoinNetworkModes } from '@shared/constants';
+import { getAddressFromOutScript } from '@shared/crypto/bitcoin/bitcoin.utils';
+import { isDefined } from '@shared/utils';
+
+import { OrdApiInscriptionTxOutput } from '@app/query/bitcoin/ordinals/ordinals-aware-utxo.query';
+
+import { PsbtInput, PsbtInputForUi } from './use-psbt-decoded-request';
+
+function getPsbtInputValue(input: PsbtInput, unsignedUtxo?: OrdApiInscriptionTxOutput) {
```
Comment (kyranjamie) [inline src/app/pages/psbt-request/components/psbt-decoded-request/psbt-decoded-request-views/build-psbt-input-for-ui.ts]: How come the latter is optional?

File: src/app/pages/psbt-request/components/psbt-decoded-request/psbt-decoded-request-views/build-psbt-input-for-ui.ts [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+  return psbtInputs.map((input, i) => {
+    const utxoAddress = getPsbtInputAddress(input, network, unsignedUtxos[i]);
+    const utxoValue = getPsbtInputValue(input, unsignedUtxos[i]);
+    return {
+      ...input,
+      address: utxoAddress,
+      value: utxoValue,
+      unsignedUtxo: unsignedUtxos[i],
+    };
+  });
```
Comment (kyranjamie) [inline src/app/pages/psbt-request/components/psbt-decoded-request/psbt-decoded-request-views/build-psbt-input-for-ui.ts]: Not sure I follow what's happening here. How does the index of the psbtInput relate to the unsignedUtxo?

File: src/app/pages/psbt-request/components/psbt-decoded-request/psbt-decoded-request-views/match-inputs-and-outputs.ts (hunk 1 above)
Comment (kyranjamie) [reply src/app/pages/psbt-request/components/psbt-decoded-request/psbt-decoded-request-views/match-inputs-and-outputs.ts]: would flag this as magic number, can we make a const?

Also, _is_ it the case an inscription is always 10k? It doesn't have to be. Is that what gamma does?

File: src/app/pages/psbt-request/components/psbt-decoded-request/psbt-decoded-request-views/match-inputs-and-outputs.ts [hunk 5]
```diff
@@ -0,0 +1,92 @@
...
+}
+
+interface MatchInputsOutputsArgs {
+  psbtInputs: PsbtInputForUi[];
+  unsignedOutputs: btc.TransactionOutputRequired[];
+}
+export function matchInputsOutputs({ psbtInputs, unsignedOutputs }: MatchInputsOutputsArgs): {
+  fee: number;
+  inputOutputPairs: InputOutputPair[];
+} {
```
Comment (kyranjamie) [inline src/app/pages/psbt-request/components/psbt-decoded-request/psbt-decoded-request-views/match-inputs-and-outputs.ts]: This is fairly complex so would encourage unit tests.

Separately, I wonder if some of this logic might better belong elsewhere. It's nested quite deep in the component tree. Maybe we could separate out higher in some general `psbt` logic folder?

File: src/app/pages/psbt-request/components/psbt-decoded-request/psbt-decoded-request-views/use-psbt-decoded-request.tsx [hunk 6]
```diff
@@ -0,0 +1,110 @@
...
+  inputs: btc.TransactionInputRequired[];
+  outputs: btc.TransactionOutputRequired[];
+  witnesses: string[][];
+  lockTime: number;
+}
+
+interface WitnessUtxo {
+  script: Uint8Array;
+  amount: bigint;
+}
```
Comment (kyranjamie) [inline src/app/pages/psbt-request/components/psbt-decoded-request/psbt-decoded-request-views/use-psbt-decoded-request.tsx]: Can we pull from the lib?

File: src/app/pages/psbt-request/components/psbt-decoded-request/psbt-decoded-request-views/use-psbt-decoded-request.tsx [hunk 7]
```diff
@@ -0,0 +1,110 @@
...
+      matchInputsOutputs({
+        psbtInputs: inputs,
+        unsignedOutputs,
+      }),
+    [inputs, unsignedOutputs]
+  );
+
+  useEffect(() => {
+    if (
+      isPlaceholderTransaction(
```
Comment (kyranjamie) [inline src/app/pages/psbt-request/components/psbt-decoded-request/psbt-decoded-request-views/use-psbt-decoded-request.tsx]: Seems odd to call this in an effect and set the state. Plus, if it were later false, would never change back?

Should we just return `isPlaceholderTransaction` in the object directly, as it's an expression?

File: src/shared/crypto/bitcoin/bitcoin.utils.ts [hunk 8]
```diff
@@ -37,23 +40,34 @@ export function decodeBitcoinTx(tx: string) {
...
-  return btc.programToWitness(1, script.slice(2));
+function getAddressFromTrOutScript(script: Uint8Array, network: BitcoinNetwork) {
+  return btc.programToWitness(1, script.slice(2), network);
 }
 
-export function getAddressFromOutScript(script: Uint8Array) {
+export function getAddressFromOutScript(script: Uint8Array, network: BitcoinNetworkModes) {
   const outScript = btc.OutScript.decode(script);
+  // This appears to be undefined at times?
+  const bitcoinNetwork = getBtcSignerLibNetworkConfigByMode(network);
```
Comment (kyranjamie) [inline src/shared/crypto/bitcoin/bitcoin.utils.ts:64]: Why is this? We have a map value for all the `BitcoinNetworkModes`?

File: src/app/pages/psbt-request/components/psbt-decoded-request/psbt-decoded-request-views/build-psbt-input-for-ui.ts (hunk 3 above)
Comment (fbwoolf) [reply src/app/pages/psbt-request/components/psbt-decoded-request/psbt-decoded-request-views/build-psbt-input-for-ui.ts]: Bc Gamma (and others) often send 'dummy' unsigned inputs in the psbt, so the ord api doesn't return data.

File: src/app/pages/psbt-request/components/psbt-decoded-request/psbt-decoded-request-views/build-psbt-input-for-ui.ts (hunk 4 above)
Comment (fbwoolf) [reply src/app/pages/psbt-request/components/psbt-decoded-request/psbt-decoded-request-views/build-psbt-input-for-ui.ts]: There are unsigned inputs and inputs/outputs containing the witness/non-witness data on the psbt, both have to be used in dep on what is passed in the psbt. The unsigned input index is the same as the input index on the psbt.

File: src/app/pages/psbt-request/components/psbt-decoded-request/psbt-decoded-request-views/match-inputs-and-outputs.ts (hunk 5 above)
Comment (fbwoolf) [reply src/app/pages/psbt-request/components/psbt-decoded-request/psbt-decoded-request-views/match-inputs-and-outputs.ts]: It is always `10000` for the sell or buy. The inscription utxo itself may not be 10000. But the new input/output is.

Comment (fbwoolf) [reply src/app/pages/psbt-request/components/psbt-decoded-request/psbt-decoded-request-views/match-inputs-and-outputs.ts]: > This is fairly complex so would encourage unit tests.
> 
> Separately, I wonder if some of this logic might better belong elsewhere. It's nested quite deep in the component tree. Maybe we could separate out higher in some general `psbt` logic folder?

Do you anticipate psbts being used elsewhere in the code? Not sure why I would move unless it will be used elsewhere?

File: src/app/pages/psbt-request/components/psbt-decoded-request/psbt-decoded-request-views/use-psbt-decoded-request.tsx (hunk 6 above)
Comment (fbwoolf) [reply src/app/pages/psbt-request/components/psbt-decoded-request/psbt-decoded-request-views/use-psbt-decoded-request.tsx]: Nothing was exported that I saw that is why these were created. I can check again, but not a lot is exported.

File: src/app/pages/psbt-request/components/psbt-decoded-request/psbt-decoded-request-views/use-psbt-decoded-request.tsx (hunk 7 above)
Comment (fbwoolf) [reply src/app/pages/psbt-request/components/psbt-decoded-request/psbt-decoded-request-views/use-psbt-decoded-request.tsx]: I thought I had to do this for the unsignedUtxos query to load, but let me check. Good point though, will change to toggle.

File: src/shared/crypto/bitcoin/bitcoin.utils.ts (hunk 8 above)
Comment (fbwoolf) [reply src/shared/crypto/bitcoin/bitcoin.utils.ts:64]: I couldn't track this down but def logged undefined and errored unless I used `?.` which now has a lint warning bc it shouldn't be undefined.

File: src/app/pages/psbt-request/components/psbt-decoded-request/psbt-decoded-request-views/build-psbt-input-for-ui.ts (hunk 4 above)
Comment (fbwoolf) [reply src/app/pages/psbt-request/components/psbt-decoded-request/psbt-decoded-request-views/build-psbt-input-for-ui.ts]: Fwiw, I think trying to handle how an app passes us a psbt and rendering it in the UI is error prone ...like trying to handle nonces. But, I have been trying to throw the red flag on it for awhile now ...which is why I put a comment on the PR that we should have something in place for when the logic doesn't appear to work and force the advanced view only (or something).

File: src/app/pages/psbt-request/components/psbt-decoded-request/psbt-decoded-request-views/match-inputs-and-outputs.ts (hunk 1 above)
Comment (fbwoolf) [reply src/app/pages/psbt-request/components/psbt-decoded-request/psbt-decoded-request-views/match-inputs-and-outputs.ts]: It is always 10000 for the sell or buy. The inscription utxo itself may not be 10000. But the new input/output is.

Comment (kyranjamie) [reply src/app/pages/psbt-request/components/psbt-decoded-request/psbt-decoded-request-views/match-inputs-and-outputs.ts]: > It is always 10000 for the sell or buy

Requests from Gamma? What if another marketplace tries to use PSBTs for inscriptions, and doesn't follow this rule? Main thinking here is that our implementation should try to avoid details specific to given apps.

Are there any ways to avoid this? Do we absolutely need this check?

To add an example: what if I make a PSBT that takes a 10k inscription-holding utxo input and has a 1 sat output and a 9999 fee?

File: src/app/pages/psbt-request/components/psbt-decoded-request/psbt-decoded-request-views/match-inputs-and-outputs.ts (hunk 5 above)
Comment (kyranjamie) [reply src/app/pages/psbt-request/components/psbt-decoded-request/psbt-decoded-request-views/match-inputs-and-outputs.ts]: Not suggesting moving away from the feature, but from the deeply nested view, something like `src/pages/psbt-request/utils`?

Not necessary, just promoting decoupling views/business logic.

File: src/app/pages/psbt-request/components/psbt-decoded-request/psbt-decoded-request-views/match-inputs-and-outputs.ts (hunk 1 above)
Comment (fbwoolf) [reply src/app/pages/psbt-request/components/psbt-decoded-request/psbt-decoded-request-views/match-inputs-and-outputs.ts]: I thought this was an inscription rule? Magic Eden does same. I thought this was built into the protocol for some reason?

Comment (fbwoolf) [reply src/app/pages/psbt-request/components/psbt-decoded-request/psbt-decoded-request-views/match-inputs-and-outputs.ts]: Anyway, I will remove it and think of something else ...maybe it was part of one of my earlier versions and it's actually not needed now. I've gone thru many refactors of this, haha.

File: src/app/pages/psbt-request/components/psbt-decoded-request/psbt-decoded-request-views/match-inputs-and-outputs.ts (hunk 5 above)
Comment (fbwoolf) [reply src/app/pages/psbt-request/components/psbt-decoded-request/psbt-decoded-request-views/match-inputs-and-outputs.ts]: Interesting, ok, I thought better to locate near where it is being used if not shared.

Comment (fbwoolf) [reply src/app/pages/psbt-request/components/psbt-decoded-request/psbt-decoded-request-views/match-inputs-and-outputs.ts]: I will move it out ...I first thought you meant out into shared utils.

Comment (fbwoolf) [reply src/app/pages/psbt-request/components/psbt-decoded-request/psbt-decoded-request-views/match-inputs-and-outputs.ts]: Made a `hooks` folder higher up, think that is best bc used there and can locate hooks together.

File: src/app/pages/psbt-request/components/psbt-decoded-request/psbt-decoded-request-views/match-inputs-and-outputs.ts (hunk 1 above)
Comment (fbwoolf) [reply src/app/pages/psbt-request/components/psbt-decoded-request/psbt-decoded-request-views/match-inputs-and-outputs.ts]: I was able to just remove this, sorry for the confusion.

File: src/app/pages/psbt-request/components/psbt-decoded-request/psbt-decoded-request-views/use-psbt-decoded-request.tsx (hunk 6 above)
Comment (fbwoolf) [reply src/app/pages/psbt-request/components/psbt-decoded-request/psbt-decoded-request-views/use-psbt-decoded-request.tsx]: Prob going to have to build out a lot more with types here to handles all possible that could be passed? Signer lib has a lot of code handling it but doesn't export to use.

**Comment by mica000** [comment]
@fbwoolf Looking good!

About the suggestion of showing the Ordinal value, wondering if it's a good idea to lose the number of the inscription tho. Made a spin of this idea covering both elements:

Also wondering about the display of the value in USD, is this too hard to be displayed?


Option 1
Screenshot 2023-04-11 at 14 27 18

Option 2
Screenshot 2023-04-11 at 14 33 20

**Comment by fbwoolf** [comment]
> About the suggestion of showing the Ordinal value, wondering if it's a good idea to lose the number of the inscription tho. Made a spin of this idea covering both elements:
> 
We can do, it is just another query bc we don't have that information. Will the image or anything link to the inscription?
>
> Also wondering about the display of the value in USD, is this too hard to be displayed?
>
I just overlooked this part, thanks for pointing out, I will add it.

**Comment by mica000** [comment]
> > About the suggestion of showing the Ordinal value, wondering if it's a good idea to lose the number of the inscription tho. Made a spin of this idea covering both elements:
> 
> We can do, it is just another query bc we don't have that information. Will the image or anything link to the inscription?
> 
> > Also wondering about the display of the value in USD, is this too hard to be displayed?
> 
> I just overlooked this part, thanks for pointing out, I will add it.


What do you think of this?
Screenshot 2023-04-11 at 16 04 29

**Comment by mica000** [comment]
@fbwoolf About the address we are displaying, wondering if we can make that by hovering so the user is able to copy it so they can always make sure it's sending/receiving in the right address

File: src/app/pages/psbt-request/components/psbt-request-header.tsx [hunk 9]
```diff
@@ -13,21 +13,21 @@ function PsbtRequestHeaderBase() {
...
-    </>
+    <Flex flexDirection="column" my="loose">
+      <Title fontSize={3} fontWeight={500} mb="base">
+        Sign transaction
+      </Title>
+      {caption && (
+        <Stack isInline pl="tight">
+          <img src={appIcon} height="16px" width="16px" />
+          <Caption wordBreak="break-word">{caption}</Caption>
+        </Stack>
```
Comment (kyranjamie) [inline src/app/pages/psbt-request/components/psbt-request-header.tsx]: Could be a flag?

File: src/app/pages/psbt-request/hooks/match-inputs-and-outputs.ts [hunk 10]
```diff
@@ -0,0 +1,87 @@
...
+      const input = availableInputs[i];
+
+      if (remainingInputValue === 0) remainingInputValue = input.value;
+
+      // Handle if the unsigned input at index has an inscription
+      if (availableInputs[i].unsignedUtxo?.inscriptions) {
+        inputOutputPair.inputs.push(input);
+        remainingInputValue = 0;
+        availableInputs.splice(i, 1);
+        break;
```
Comment (kyranjamie) [inline src/app/pages/psbt-request/hooks/match-inputs-and-outputs.ts]: Why do we break the loop entirely when there's potentially more inputs that'll be signed? 🤔

File: src/app/pages/psbt-request/hooks/mock-psbts.ts [hunk 11]
```diff
@@ -0,0 +1,257 @@
+import { hexToBytes } from '@stacks/common';
```
Comment (kyranjamie) [inline src/app/pages/psbt-request/hooks/mock-psbts.ts]: Would include in test file or something, seems weird to have mock data not in a test. Or move elsewhere?

**Review by kyranjamie — APPROVED** [review]

File: src/app/pages/psbt-request/hooks/match-inputs-and-outputs.ts (hunk 10 above)
Comment (fbwoolf) [reply src/app/pages/psbt-request/hooks/match-inputs-and-outputs.ts]: It just breaks it for that output bc it has been fulfilled, but continues with other pairing. I have only seen one input to an output for an inscription, is that not always the case?

File: src/app/pages/psbt-request/hooks/mock-psbts.ts (hunk 11 above)
Comment (fbwoolf) [reply src/app/pages/psbt-request/hooks/mock-psbts.ts]: Ok, yeah, just seemed like a lot to include in the test file but will move it if preferred.

File: src/app/pages/psbt-request/hooks/match-inputs-and-outputs.ts (hunk 10 above)
Comment (kyranjamie) [reply src/app/pages/psbt-request/hooks/match-inputs-and-outputs.ts]: Maybe a comment here would help explain what's going on, I'm not really sure what it does. 

Input/outputs can be many-to-one, one-to-many, many-to-many, and referring to them as "pairs" doesn't align with that.

I think the complexity of this feature lies in the blur between building a "psbt signing feature" vs "inscription marketplace transfers (that just happens to use psbts)". Ideally, inscription-specific features are somehow layered on top of an inscription-agnostic psbt imlpementation.

Comment (fbwoolf) [reply src/app/pages/psbt-request/hooks/match-inputs-and-outputs.ts]: Yes, I agree, but how do I confront that with an issue/designs specifically targeting inscription transfers? To me, this would be decided before it gets to development, which I tried to point out several times. @markmhx 

I feel like what I have done does handle the many-to-one and one-to-many you have described bc the pair has inputs as an array, however I'm not sure my implementation handles many-to-many. I feel like I'm trying to write code to predict how the blockchain will handle transactions, much like we try to write logic to predict nonces. It is error prone. I will think thru the many-to-many.

Comment (fbwoolf) [reply src/app/pages/psbt-request/hooks/match-inputs-and-outputs.ts]: And, are there any examples that anyone else is even doing this ...or that it is requested?

Comment (markmhendrickson) [reply src/app/pages/psbt-request/hooks/match-inputs-and-outputs.ts]: How would this currently handle the many-to-many scenario in which you have 1 input from address X that produces outputs to addresses A,B,C, etc.?

Presumably we'd see a 10 groupings of "Output/Input" with the input data shown in a redundant way? Or would something else happen?

If the former, it may not be the most ideal UX but seems viable for an iterative enhancement here.

Comment (fbwoolf) [reply src/app/pages/psbt-request/hooks/match-inputs-and-outputs.ts]: Yeah, after thinking on this, I am handling many-to-many bc the input value is used across multiple outputs. You will just see the input address as the same for multiple outputs. @kyranjamie have you run this branch/tested it to see it in the UI? This might just be that I misunderstood @kyranjamie's comment and that I should just not handle inscriptions in the matching, but only match based on value ...then handle if an inscription is present later? Also, it appears I just need to add comments in the code to better dictate what is happening bc it does appear to be working in the tests I have done so far.

Comment (kyranjamie) [reply src/app/pages/psbt-request/hooks/match-inputs-and-outputs.ts]: Comment related mostly to:
1) why the break
2) why inscription handling within pairing

Happy to chat through further if needed, as long as impl. works with generic non-inscription tx that's fine

Comment (fbwoolf) [reply src/app/pages/psbt-request/hooks/match-inputs-and-outputs.ts]: I just removed handling the inscription in the matching and seems to be ok, right? Now multiple inputs cover the output...

Screen Shot 2023-04-12 at 12 16 47 PM

Comment (fbwoolf) [reply src/app/pages/psbt-request/hooks/match-inputs-and-outputs.ts]: This was always working when an input wasn't an inscription, but I had overlooked would be the same with an inscription.

**Comment by fbwoolf** [comment]
@mica000 are you ok with the wrap here, the taproot addresses can get quite long and overflow...

Screen Shot 2023-04-12 at 12 59 27 PM
